src/engine.rs

```diff
@@ -6,73 +6,133 @@ use crate::interpreter::*;
 use crate::lexer::*;
 use crate::parser::*;
 use crate::scheduler::*;
+use crate::utils::gather_functions;
 use crate::value::*;
 
 use anyhow::Result;
 
 #[derive(Clone)]
 pub struct Engine {
     modules: Vec<Ref<Module>>,
-    input: Value,
-    data: Value,
-}
-
-impl Default for Engine {
-    fn default() -> Self {
-        Self::new()
-    }
+    interpreter: Interpreter,
+    prepared: bool,
 }
 
 impl Engine {
-    pub fn new() -> Self {
-        Self {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
             modules: vec![],
-            input: Value::new_object(),
-            data: Value::new_object(),
-        }
+            interpreter: Interpreter::new()?,
+            prepared: false,
+        })
     }
 
     pub fn add_policy(&mut self, path: String, rego: String) -> Result<()> {
         let source = Source::new(path, rego);
         let mut parser = Parser::new(&source)?;
         self.modules.push(Ref::new(parser.parse()?));
+        // if policies change, interpreter needs to be prepared again
+        self.prepared = false;
         Ok(())
     }
 
     pub fn add_policy_from_file(&mut self, path: String) -> Result<()> {
         let source = Source::from_file(path)?;
         let mut parser = Parser::new(&source)?;
         self.modules.push(Ref::new(parser.parse()?));
+        self.prepared = false;
         Ok(())
     }
 
     pub fn set_input(&mut self, input: Value) {
-        self.input = input;
+        self.interpreter.set_input(input);
     }
 
     pub fn clear_data(&mut self) {
-        self.data = Value::new_object();
+        self.interpreter.set_data(Value::new_object());
+        self.prepared = false;
     }
 
     pub fn add_data(&mut self, data: Value) -> Result<()> {
-        self.data.merge(data)
+        self.prepared = false;
+        self.interpreter.get_data().merge(data)
     }
 
-    pub fn eval_query(&self, query: String, enable_tracing: bool) -> Result<QueryResults> {
-        // Analyze the modules and determine how statements must be scheduled.
-        let analyzer = Analyzer::new();
-        let schedule = analyzer.analyze(&self.modules)?;
+    pub fn prepare_for_eval(&mut self, enable_tracing: bool) -> Result<()> {
```

Comment:
Does this function need to be public?

Reply:
I think not, previously it was necessary for the end-user to explicitly calls this function and now it's done automatically behind the hood when calling `eval_*` functions of the engine. I don't see a valid reason for the end-user to call it.